# 异常处理与IO操作

异常不是语法错误，他不是由程序员的疏忽造成的，而是一套保护机制

## 错误与异常

都会打断正常的运行流程，都是Throwable的子类

错误是靠程序代码本身有效的恢复，常见的错误有内存不足、方法调用栈溢出

### 异常处理的固定方式，try...catch...finally语句

```java
try{
    需要监控的可能会抛出异常的代码块
}
catch(Exception e){
    出现异常后的处理代码
      //System.exit(0);  执行之后不会执行finally
}finally{
    回收资源的动作
}
```

在catch代码块中可以捕捉异常并用合理的方法来处理该异常

- 运行期异常类不必包含在try从句中

### throw ，throws的区别

- throws 出现在声明方法的位置，而throw 出现在函数体中。
- 如果在某个函数内部用throw 抛出异常，那么在声明该函数时，一定要配套使用throws ，否则会出现语法错误
- 如果在某个函数的声明位置出现了用throws 抛出异常的情况，那么就需要用try...catch代码块来包含调用的代码，否则也会出现语法错误
- 子类方法中不应该抛出比父类范围更广的异常

## return两个作用

- 返回方法制定类型的值
- 结束方法的执行（仅仅一个return;）